fleetspeak/src/e2etesting/lib/setup_components.go

```
@@ -0,0 +1,296 @@
+	}
+
+	stdoutIn, _ := cmd.StdoutPipe()
+	stderrIn, _ := cmd.StderrPipe()
+	err := cmd.Start()
+	if err != nil {
+		fmt.Println("cmd.Start() failed: " + err.Error())
+	}
+	pid_chan <- cmd.Process.Pid
+
+	// cmd.Wait() should be called only after we finish reading
+	// from stdoutIn and stderrIn.
+	// wg ensures that we finish
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		CopyAndCapture(os.Stdout, stdoutIn)
+		wg.Done()
+	}()
+
+	CopyAndCapture(os.Stderr, stderrIn)
+	wg.Wait()
+	cmd.Wait()
+}
+
+func getNewClientId(admin_port int, start_time time.Time) string {
+	admin_addr := fmt.Sprintf("localhost:%v", admin_port)
+	conn, err := grpc.Dial(admin_addr, grpc.WithInsecure())
+	if err != nil {
+		fmt.Printf("Unable to connect to fleetspeak admin interface [%v]: %v", admin_addr, err)
+	}
+	admin := services_pb.NewAdminClient(conn)
+	ctx := context.Background()
+	var ids [][]byte
+	for i := 1; i <= 10; i++ {
+		if i > 1 {
+			time.Sleep(time.Second * 1)
+		}
+		res, err := admin.ListClients(ctx,
+			&services_pb.ListClientsRequest{ClientIds: ids},
+			grpc.MaxCallRecvMsgSize(1024*1024*1024))
+		if err != nil {
+			fmt.Printf("ListClients RPC failed: %v", err)
+			continue
+		}
+		if len(res.Clients) == 0 {
+			fmt.Printf("No new clients yet")
+			continue
+		}
+		client := res.Clients[0]
+		last_visit_time, _ := ptypes.Timestamp(client.LastContactTime)
+		for _, cl := range res.Clients {
+			cur_last_visit_time, _ := ptypes.Timestamp(cl.LastContactTime)
+			if cur_last_visit_time.After(last_visit_time) {
+				client = cl
+				last_visit_time = cur_last_visit_time
+			}
+		}
+
+		id, err := common.BytesToClientID(client.ClientId)
+		if err != nil {
+			fmt.Printf("Ignoring invalid client id [%v], %v", client.ClientId, err)
+			continue
+		}
+		ts, err := ptypes.Timestamp(client.LastContactTime)
+		if ts.After(start_time) {
+			return fmt.Sprintf("%v", id)
+		}
+	}
+	return ""
+}
+
+func configureFleetspeak(temp_path string, fs_frontend_port, fs_admin_port int) {
```

Please make this function return an error and propagate all errors to the caller.

As a general rule:
- We **never** ignore errors.
- We almost never print errors, we always propagate them to the caller and ideally, if we do print them, we do it in main()